### DISCUSSION TITLE
OsmAnd Live data overrides custom-installed OBF file even when the corresponding offline map is deactivated

### DISCUSSION THREAD

**Discussion by mattmccutchen** [osmandapp/OsmAnd#13543] [Q&A]
### Description
I often take advantage of the ability to install and view custom OBF files in OsmAnd to validate my OSM changesets before submission.  My workflow looks like this:

1. Draft a changeset in JOSM and save it as an OSM file.
2. Convert the OSM file to an OBF file using OsmAndMapCreator following [these instructions](https://docs.osmand.net/en/main@latest/development/map-creation/create-offline-maps-yourself#vector-maps-shell-script).
3. Deactivate the OsmAnd-provided monthly offline map of the region.  Copy my OBF file into the deactivated (`backup`) directory and then activate it.
4. Inspect what I changed on the map.
5. Return OsmAnd to the original state by deleting my OBF file and reactivating the monthly map.

If I don't deactivate the monthly map in step 3, then I see some mix of the OsmAnd-provided data and my custom data and I cannot validate my changeset, which is understandable.  Until now, I assumed that if I deactivated the monthly map, I would always see my custom data.  However, today I saw a case in which OsmAnd Live data I previously downloaded seemed to override my custom OBF file, even though the corresponding monthly map was deactivated and the region was no longer visible in the OsmAnd Live window at all.  This behavior is very confusing.  I'd expect OsmAnd Live data to be used only when a corresponding monthly map is installed and active.

As a workaround, I'm now validating my changesets using a second instance of OsmAnd with no OsmAnd-provided maps installed whatsoever.

### How to reproduce?
I'll give a specific example that works as of this writing (2022-01-09) based on a change I recently submitted that is currently in OsmAnd Live.  I presume it will stop working once that change gets incorporated into the 2021-12-31 monthly map, but it should be possible to find another example at that point.  Here's [a zip file](https://github.com/osmandapp/OsmAnd/files/7835864/live-override-test-20220109.zip) containing the custom OSM and OBF files for testing.
1. Ensure that Us_maryland_northamerica.obf dated 2021-11-30 is installed.  Go to the OsmAnd Live window and ensure that any OsmAnd Live data for Maryland is deleted.  View [this node for a Giant supermarket](https://www.openstreetmap.org/node/558678196) in OsmAnd.  There should be no URL, which is correct: that was the state of OpenStreetMap as of 2021-11-30.
2. Deactivate Us_maryland_northamerica.obf.  Install and activate live-override-test-20220109.obf from the zip file linked above.  View the Giant again.  The URL should start with `https://TEST.stores.giantfood.com/`, which is correct: that's from the custom OSM file.
3. Deactivate live-override-test-20220109.obf and reactivate Us_maryland_northamerica.obf.  Now install the latest OsmAnd Live data for Maryland.  View the Giant again.  The URL should start with `https://stores.giantfood.com/`, which is correct: that's the current state of OpenStreetMap.
4. Deactivate Us_maryland_northamerica.obf and reactivate live-override-test-20220109.obf.  View the Giant again.  The URL starts with `https://stores.giantfood.com/`, which is unexpected; I assume it's coming from the OsmAnd Live data.  It should start with `https://TEST.stores.giantfood.com/` as in the custom OSM file.

### Your Environment
OsmAnd Version: 4.0.9 from F-Droid.  (I can't easily test the latest nightly build because I don't see how I can enable OsmAnd Live there without paying.)

Android/iOS version: Android 11, LineageOS 18.1-20211216-NIGHTLY-sunfish

Device model: Google Pixel 4a

**Maps used (online or offline):**   
If you have an issue related to offline maps, tell us the exact name of the map file where the issue occurs and its edition date.

Offline Us_maryland_northamerica.obf dated 2021-11-30

**Comment by vshcherb** [comment]
Hm... that's a bit sophisticated I would do it simpler... just always create a custom changeset for next year or next month
i.e.
```
2021_01_00 - month
2021_01_01 - today
2021_12_31 - your today changeset activate / deactivate
```

Please remind that you can't create a correct changeset file without Overpass! (There is youtube presentation about details I gave). Without overpass you would have wrong relations / wrong turn restrictions and sometimes wrong geometry. 

Of course if you generate OSM for the whole area that's not a problem but still.. Changeset not only creates new entities but properly generates old entities with your approach you will just double entities and deleted entities won't be hidden

**Reply by mattmccutchen** [reply]
Indeed, I'm generating OBF for the whole (small) area that I downloaded in JOSM and (trying to) deactivate all other data for the same region in OsmAnd.  Then in that region, I should should see only my OBF file, without duplicate or deleted entities, right?

The idea to create a fake OsmAnd Live delta file is clever; I didn't think of that.  However, there are several reasons why I feel that my approach is simpler:

1. The custom OBF file shows up on the "Local" tab of the "Download maps" screen, so I can confirm that I installed it in the correct folder and activate it or deactivate it from the UI.  When I activate or deactivate an OBF file in the UI, the map view seems to reliably update right away.  By comparison, OsmAnd Live feels like a black box and I don't feel confident to troubleshoot any problems.  I don't see any UI to list, activate, or deactivate individual OsmAnd Live delta files; I assume I would have to manually add or remove files in the `live` directory.  If I do that while OsmAnd is running, the map view doesn't update right away.  What am I supposed to do to make it update?  Force-stop and restart OsmAnd?
2. To generate the OBF file for the whole area, I just follow [these instructions](https://docs.osmand.net/en/main@latest/development/map-creation/create-offline-maps-yourself#vector-maps-shell-script).  What's the procedure to generate an OsmAnd Live delta file for a change that I haven't uploaded yet?  Is there any documentation?  I assume it wouldn't make sense to use a public Overpass server because the server only knows about changes that have been uploaded.
3. When I use an OBF file for the whole area that I saved from JOSM, I'm guaranteed that I'm seeing the same data for the whole area in OsmAnd as in JOSM.  That data might be out of date if other changesets have been uploaded after I initially downloaded the data in JOSM, but at least it's an internally consistent snapshot.  If I create a custom OsmAnd Live delta file from my change, then the "old" state of the custom delta (based on the time I downloaded the data in JOSM) may not match the "new" state of the latest official delta file from OsmAnd Live.  In this situation, can problems occur when OsmAnd tries to apply the custom delta?  Even if not, it could be confusing if the elements I modified reflect some previously uploaded changesets that the other elements don't, or vice versa.

Unless/until these issues are satisfactorily addressed, I'll continue to use and [recommend](https://wiki.openstreetmap.org/wiki/Testing_changes_before_upload#OsmAnd) the use of custom OBF files for the whole area.  So I think my issue report about OsmAnd using Live data for a deactivated map has merit as originally written.  However, using a second OsmAnd instance for the custom data may be a good idea for other reasons in addition to avoiding the issue with OsmAnd Live, in which case solving that issue would be lower priority.  Specifically: Sometimes I find it helpful to compare the map view before and after my change, and I find it easier to do that by switching between the apps with the Android "recent apps" button than by activating and deactivating files.  (Ideally, I could even use Android's split-screen mode to see both apps at once.  Currently, when I try that, sometimes OsmAnd gets confused and gives me a black screen, but that's a topic for a separate issue report.)

> There is youtube presentation about details I gave

For the record, I assume you mean [this one](https://www.youtube.com/watch?v=w0AU3ARMeok)?  It's interesting and good for me to know in general but doesn't seem to address any of my questions above.

**Reply by thompehw** [reply]
AFAIK when there are objects with the same ID in multiple map/update files, it's the one from the file that comes last in filename sort order that is used, and it doesn't matter whether the files are in the "live" folder or in with the main maps.

> Deactivate Us_maryland_northamerica.obf and reactivate live-override-test-20220109.obf. View the Giant again. The URL starts with https://stores.giantfood.com/, which is unexpected; I assume it's coming from the OsmAnd Live data. It should start with https://TEST.stores.giantfood.com/ as in the custom OSM file.

If I'm right, this is happening because the live update files all begin "Us" so are later in sort order than "live-override..."

I have done the same approach in the past and just called my update file "zzz.obf" - easier than adding a date suffix!

BTW I discovered that by adding the tag "osmand_change=delete" to an object your update file can delete it, but as this is not an OSM tag it's of limited use in testing a potential changeset.

**Reply by vshcherb** [reply]
Yes presentation is correct :-)

**Reply by vshcherb** [reply]
With custom OBF files it's not possible to mark routes as deleted, so it's working good but not complete
